ci/unit_tests/functions_test/main_test.py

```diff
@@ -0,0 +1,268 @@
+            1, # exit code
+            'CRITICAL Cannot decode json from test input file ' + 
+            '/home/tomas/Projects/watson-assistant-workbench/ci/unit_tests/functions_test/main_data/no.json, ' + 
+            'error: No JSON object could be decoded', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_nonArrayJsonInput(self):
+        ''' Tests if the input file contains non array json '''
+        testArgs = [self.emptyDictJsonPath, self.noJsonPath] + self.functionsTestArgs
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'CRITICAL Input test json is not array!', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_testSingleInvalid(self):
+        ''' Tests if the input file contains test that is not dictionary '''
+        testArgs = [self.testSingleInvalidJsonPath, self.noJsonPath] + self.functionsTestArgs
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'CRITICAL Input test array element 0 is not dictionary. Each test has to be dictionary, please see doc!', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+
+    def test_testSingleAllInFileBadAuthentification(self):
+        ''' Tests if bad authentification was provided '''
+        testArgs = [self.testSingleAllInFileJsonPath, self.noJsonPath] + self.functionsTestArgs
+        testArgs[testArgs.index('--cloudfunctions_password') + 1] = 'invalidPassword' # change password to something different
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            'The supplied authentication is invalid', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_testSingleAllInFileBadUrl(self):
+        ''' Tests if bad url was provided '''
+        testArgs = [self.testSingleAllInFileJsonPath, self.noJsonPath] + self.functionsTestArgs
+        testArgs[testArgs.index('--cloudfunctions_url') + 1] = 'https://us-south.functions.cloud.ibm.com/invalidUrl' # change password to something different
+        self.t_exitCodeAndLogMessage(
+            1, # exit code
+            '404 Not Found', # error message substring
+            [testArgs] # params (*args, **kwargs)
+        )
+
+    def test_testSingleAllInFile(self):
+        ''' Tests if the single test where input and expected output is specified in file (all other params are given from command line) '''
+        outputFilePath = os.path.abspath(os.path.join(self.testOutputPath, os.path.splitext(os.path.basename(self.testSingleAllInFileJsonPath))[0] + '.out.json'))
+        testArgs = [self.testSingleAllInFileJsonPath, outputFilePath] + self.functionsTestArgs
+        self.t_noException([testArgs])
+        with open(outputFilePath, 'r') as outputFile:
+            outputJson = json.load(outputFile)
+            assert outputJson[0]['result'] == 0
```

Comment:
1. You should check first if there is an array in `outputJson` and it has exactly one item
2. Consider adding `assert` which checks that there is no attribute `diff`

Reply:
I'll improve it.